### DISCUSSION TITLE
Suppressing negative massfracs

### DISCUSSION THREAD

**Discussion by RSuryaNarayan** [AMReX-Combustion/PeleC#764] [Q&A] [answered]
Hello, 

Is there a way to suppress negative massfractions and reset them to 0? I do know that I could do this in ```post_restart``` using the ```state```, but I am wondering if there's something already inside Pele that does this. Currently, this is causing CVODE to crash...

I observe this with methane combustion particularly. H2 with Li-Dryer works just fine!

**Comment by marchdf** [comment, marked as answer]
Hi, we currently do not have an option for that and have been very careful about introducing such an option. We let the user do that for the EB because the current state redistribution is not designed for algebraic constraints among variables. Typically, we've found that negative mass fractions away from the EB are a sign of something else going wrong. However, if you feel like this would solve your problem and that you can live with the consequences, then doing something in `post_advance` to clean the mass fractions after each time step, would work. Note that if you do that, you have to worry about mass conservation, consistency among other variables, etc.

**Reply by RSuryaNarayan** [reply]
Thanks @marchdf. I don't have EBs at all in my domain and for some reason the massfracs start getting negative in the order of (1e-24) or so and slowly build up to a large negative value (O(1e-5) to O(1e-3))

**Reply by marchdf** [reply]
The code can typically survive such small negative mass fractions. More concerning is that they become so large. Is there a pathological cell somewhere in the domain? At an inlet/wall/outlet?

**Reply by RSuryaNarayan** [reply]
yes these are like just a bunch of cells in the domain. I am trying to simulate methane/O2/N2 detonations in the shock frame using a reduced FFCM mechanism in this [paper](https://www.sciencedirect.com/science/article/abs/pii/S0010218024000890). Temperature looks like this:

<img width="507" alt="image" src="https://github.com/AMReX-Combustion/PeleC/assets/52670594/50d32d6a-cab1-4df6-ad43-a945a39bdf6d"> 

There are only few cells where CH2O massfraction goes to negative values in the order of 1e-5 like this: 

<img width="507" alt="image" src="https://github.com/AMReX-Combustion/PeleC/assets/52670594/e7fdb323-b30f-4589-a4a8-0235120d5ef7">

I've give the case a thorough check...nothing immediately wrong I feel. Can email you too to have a look if needed. Appreciate your help!

**Reply by baperry2** [reply]
Besides EB issues, another potential cause of negative mass fractions is doing the chemistry solve with GMRES and not using sufficiently tight tolerances, which would allow numerical error to build up. If you're running with `cvode.solve_type = "GMRES"`, try reducing `ode.rtol` and `ode.atol` and see if this reduces the magnitude of the negatives. Note this could make your simulation substantially slower. Using one of the direct solvers for `cvode.solve_type` (options vary depending on hardware) might also be more robust than `GMRES`.

**Reply by RSuryaNarayan** [reply]
Thanks @baperry2 this is indeed a direction I thought about. For that run, I used ```ode.atol=1e-20```. Not sure if that is tight enough. I will try other CVODE solve options!

**Reply by baperry2** [reply]
`ode.atol=1e-20` should be plenty tight enough for GMRES and potentially could be quite slow. I would have tried rtol of 1e-12 and atol of 1e-14. But it would still worth trying a different solve type just in case.

**Reply by drummerdoc** [reply]
Years ago I hit the annoying issue of a built-in instability as follows: take the reaction O + O <=> O2, and assume you start with a small amount of O in a cell.  If the time step is even slightly too large so that you deplete the O2 to a teeny negative amount, even 1.e-30 and even after many careful steps together there, things go bad quickly.  The forward reaction goes as [O]^2, which then quickly drives [O] very negative. This one is obviously a bad actor if it's in your set, but there are likely others that have the same effect indirectly.   If you can isolate a cell that goes bad, you can look for that and we can offer suggestions around it.

**Reply by RSuryaNarayan** [reply]
thanks @drummerdoc! Interestingly, I find these oscillations to get smaller (O(1e-7) i.e. 0.1ppm) when I switch to the ```FFCM1_Red``` mechanism already there in the PelePhysics repo. Just that I can't find a citation for that anywhere but it looks to work well with ```ReactorRK64```

although, is there a way to know when CVODE crashes, which exact cell(s) are the problematic ones leading to the crash?

**Reply by drummerdoc** [reply]
More interestingly, Hai Wang, the keeper of FFCM1 doesn't recognize the Red mech at all, but did send me an updated reduced mech that I will insert into PelePhysics when I get a chance.  Since the new one will be referenced and supported, it would be best if that worked for you as a better replacement.  I'll let you know when I do that.

**Reply by RSuryaNarayan** [reply]
@drummerdoc thank you for the information. If you are talking about the 12 species one (FFCMy_12) mechanism here: https://www.sciencedirect.com/science/article/pii/S0010218024000890 have it imported into PelePhysics already. Will port FFCM1-21 mentioned in the paper also into PeleC and see if that works. Thanks a lot!
